emulator.c

```diff
@@ -5,10 +5,279 @@
+  update_sign_flag(cpu, *reg);
+  update_parity_flag(cpu, *reg);
+  return 5; // NOLINT
+}
+
+// Increment Register Pair
+int
+INX(i8080 *cpu, int pair)
+{
+  uint16_t value = readRegisterPair(cpu, pair) + 1;
+  writeRegisterPair(cpu, pair, value);
+  return 5; // NOLINT
+}
+
+// Jump to Address
+int
+JMP(i8080 *cpu)
+{
+  writeRegisterPair(cpu, PC, getImmediate16BitValue(cpu));
+  return 10; // NOLINT
+}
+
+// Load Accumulator
+int
+LDAX(i8080 *cpu, int pair)
+{
+  uint16_t address = readRegisterPair(cpu, pair);
+  cpu->a = cpu_read_mem(cpu, address);
+  return 7; // NOLINT
+}
+
+// Load 16-bit Data to Register Pair
+int
+LXI(i8080 *cpu, int pair, uint16_t value)
+{
+  writeRegisterPair(cpu, pair, value);
+  return 10; // NOLINT
+}
+
+// Move register value
+int
+MOV(uint8_t *dest, const uint8_t *src)
+{
+  *dest = *src;
+  return 5; // NOLINT
+}
+
+// Move register value to memory
+int
+MOV_TO_MEM(i8080 *cpu, const uint8_t *reg)
+{
+  cpu_write_mem(cpu, readRegisterPair(cpu, HL), *reg);
+  return 7; // NOLINT
+}
+
+// Move memory to register
+int
+MOV_FROM_MEM(i8080 *cpu, uint8_t *reg)
+{
+  *reg = cpu_read_mem(cpu, readRegisterPair(cpu, HL));
+  return 7; // NOLINT
+}
+
+// Move 8-bit Data to Register
+int
+MVI(uint8_t *reg, uint8_t value)
+{
+  *reg = value;
+  return 7; // NOLINT
+}
+
+// No Operation
+int
+NOP()
+{
+  return 4; // NOLINT
+}
+
+// Pop from Stack
+int
+POP(i8080 *cpu, int pair)
+{
+  uint16_t value = ((cpu_read_mem(cpu, cpu->sp + 1) << BYTE)
+                    | cpu_read_mem(cpu, cpu->sp));
+  writeRegisterPair(cpu, pair, value);
+  cpu->sp += 2;
+  return 10; // NOLINT
+}
+
+// Push to Stack
+int
+PUSH(i8080 *cpu, int pair)
+{
+  uint16_t value = readRegisterPair(cpu, pair);
+  cpu_write_mem(cpu, cpu->sp - 1,
+                (uint8_t)((value & UPPER_8_BIT_MASK) >> BYTE));
+  cpu_write_mem(cpu, cpu->sp - 2, (uint8_t)(value & LOWER_8_BIT_MASK));
+  cpu->sp -= 2;
+  return 11; // NOLINT
+}
+
+// Return to Address
+int
+RET(i8080 *cpu)
+{
+  uint16_t address = ((cpu_read_mem(cpu, cpu->sp + 1) << BYTE)
+                      | cpu_read_mem(cpu, cpu->sp));
+  cpu->sp += 2;
+  writeRegisterPair(cpu, PC, address);
+  return 10; // NOLINT
+}
+
+// Logical XOR with Accumulator
+int
+XRA(i8080 *cpu, const uint8_t *reg)
+{
+  cpu->a = cpu->a ^ *reg;
+  update_zero_flag(cpu, cpu->a);
+  update_sign_flag(cpu, cpu->a);
+  update_parity_flag(cpu, cpu->a);
+  update_aux_carry_flag(cpu, cpu->a, MAX_8_BIT_VALUE);
+  update_carry_flag(cpu, false);
+  return 4; // NOLINT
+}
+
+// Get next byte in program counter
+uint8_t
+getImmediate8BitValue(i8080 *cpu)
+{
+  return cpu_read_mem(cpu, cpu->pc + 1);
+}
+
+// Get next two bytes in program counter
+uint16_t
+getImmediate16BitValue(i8080 *cpu)
+{
+  uint8_t lo = cpu_read_mem(cpu, cpu->pc + 1);
+  uint8_t hi = cpu_read_mem(cpu, cpu->pc + 2);
+  return (uint16_t)((hi << BYTE) | lo);
+}
+
+// Get 16-bit value for given register pair
+uint16_t
+readRegisterPair(i8080 *cpu, int pair)
```

Comment:
So this takes one of the reg pair macros, and returns the actual 16 bit int?

Reply:
Exactly. If B was 0xAA and C was 0xBB then readRegisterPair(cpu, BC) would give you 0xAABB as a 16-bit value.